### ISSUE TITLE
onSubmit not called anymore in 5.1

### ISSUE THREAD

**Issue by garenor** [aldeed/meteor-autoform#829]
I upgraded to 5.1.0 and all of a sudden none of my forms are working anymore...

I have the following setup:

````
{{#autoForm schema="shipSchema.exNames" doc=ship id="updateShipExNamesForm"}}
````

and a formhandler:

````
AutoForm.addHooks('updateShipExNamesForm', {
    onSubmit: function (insertDoc, updateDoc, currentDoc) {
        this.event.preventDefault();
        console.log('before update call exnames');
         
        ...
    
        return false;
    },
});
````

All my update forms are the same way, none are working in 5.1.0. No problem in 5.0.3. Any ideas?

**Comment by abecks** [comment]
Are you getting validation errors? I'm now getting:

```
SimpleSchema invalid keys for "stepViewForm" context:
[Object, Object]
0: Objectname: "insertDoc"type: "keyNotInSchema"value: Object__proto__: Object
1: Objectname: "updateDoc"type: "keyNotInSchema"value: Object__proto__: Objectlength: 2__proto__: Array[0]
```

It's trying to validate against insertDoc and updateDoc, instead of the contents of either. I'll see if I can track it down but probably need @aldeed for this one.

**Comment by garenor** [comment]
No, I also do not get any validation errors on most forms. It's just plain not working anymore. I only get a validation message on one form, while the provided input in the formfield is ok (works ok with 5.0.3).

**Comment by abecks** [comment]
Interesting, I'm getting validation errors for forms that do not have a type set.

Can you paste this into your project somewhere and see if they start working (only ones without a type set should be affected):

```
AutoForm.addFormType('normal', {
  onSubmit: function () {
    var c = this;

    // Get onSubmit hooks
    // These are called differently from the before hooks because
    // they run async, but they can run in parallel and we need the
    // result of all of them immediately because they can return
    // false to stop normal form submission.
    var hooks = Hooks.getHooks(this.formId, 'onSubmit');

    var hookCount = hooks.length, doneCount = 0, submitError, submitResult;

    if (hookCount === 0) {
      // we haven't called preventDefault, so normal browser
      // submission will now happen
      this.endSubmission();
      return;
    }

    // Set up onSubmit hook context
    var ctx = _.extend({
      done: function (error, result) {
        doneCount++;
        if (!submitError && error) {
          submitError = error;
        }
        if (!submitResult && result) {
          submitResult = result;
        }
        if (doneCount === hookCount) {
          // run onError, onSuccess, endSubmit
          c.result(submitError, submitResult);
        }
      }
    }, this.hookContext);

    // Call all hooks at once.
    // Pass both types of doc plus the doc attached to the form.
    // If any return false, we stop normal submission, but we don't
    // run onError, onSuccess, endSubmit hooks until they all call this.done().
    var shouldStop = false;
    _.each(hooks, function eachOnSubmit(hook) {
      var result = hook.call(ctx, c.insertDoc, c.updateDoc, c.currentDoc);
      if (shouldStop === false && result === false) {
        shouldStop = true;
      }
    });
    if (shouldStop) {
      this.event.preventDefault();
      this.event.stopPropagation();
      this.endSubmission();
    }
  },
  needsModifierAndDoc: true,
  validateForm: function () {
    // Get SimpleSchema
    var ss = AutoForm.getFormSchema(this.form.id);
    // Validate
    return AutoForm._validateFormDoc(this.formDoc.insertDoc, false, this.form.id, ss, this.form);
  }
});
```

**Comment by djhi** [comment]
@abecks got the same thing, I reverted to 5.0.3 for now

**Comment by djhi** [comment]
@abecks This is happening for forms without type, as you said. I thought default was `normal`. I'll try that tomorrow

**Comment by abecks** [comment]
@djhi the code above fixed it for me. You can paste it into your project somewhere until the repo has been updated. I'll make a PR and see what @aldeed thinks about using insertDoc for normal form validation.

**Comment by garenor** [comment]
It is reloading my page on all forms now. Validations are working fine with provided code.

**Comment by abecks** [comment]
@garenor Anyway you can give me a reproducable example? Mine are working totally fine with no type and and onSubmit hook.

Would you mind adding ```resetonSuccess=false``` to your autoForm block like this:
```
{{#autoForm id="stepViewForm" schema=schema doc=answers resetOnSuccess=false}}
```

I'm curious to know if that might be the difference between you and I.

**Comment by garenor** [comment]
Still not working. The onSubmit hook is not called either.

**Comment by abecks** [comment]
@garenor You must have something else going on. Can you share more of your code?

This is working perfectly fine:

```
<template name="tester">
    {{#autoForm schema=schema id='testForm'}}
        {{#each afFieldNames}}
            {{> afQuickField name=name options=afOptionsFromSchema}}
        {{/each}}
        <input type="submit" value="Submit">
    {{/autoForm}}
</template>
```

```
AutoForm.addHooks('testForm', {
  onSubmit: function(){
    this.event.preventDefault();
    console.log('onSubmit called');
    return false;
  }
});

Template.tester.helpers({
  schema: function(){
    return new SimpleSchema({
      test: {
        type: String
      }
    });
  }
});
```

**Comment by garenor** [comment]
Also this example is reloading my page. I don't understand why. It was working fine until now. I'll do some more checking on what might be the couse

**Comment by abecks** [comment]
Is it doing a straight up browser form submission to the current page? Do you see the form fields serialized in the URL after?

**Comment by garenor** [comment]
I should have been more clear, yes, it is doing exactly that.

**Comment by abecks** [comment]
I'm not sure if this will work but can you try this:

```
Template.autoForm.events({
  'submit form': function autoFormSubmitHandler(event, template) {
      alert('Form submission event');
  }
});
```

I'm curious to know if the form submit handler is even firing for you. I would track it down based on that though, in autoform-events.js on line 89. Easiest way is to checkout this repo to your ```packages``` directory and add debugging along the way to see where it's failing.

**Comment by coniel** [comment]
Hi, having the exact same issue as @garenor. Didn't get any alert with your code above.

**Comment by garenor** [comment]
In the above code you provided, something is going wrong in `var hooks = Hooks.getHooks(this.formId, 'onSubmit');`I quess. If I place a console.log there, it is never called.

**Comment by abecks** [comment]
Hey @coniel, have you applied my fix from above? 

It looks like autoForm's main submit handler is not firing when the form is submitted. What browser you guys using?

**Comment by garenor** [comment]
Firefox (because of firebug). Chrome is same thing, BUT, at lease chrome is giving me an error: Hooks is not defined.

**Comment by coniel** [comment]
I applied the fix and I'm using chrome. When I put a console.log the autoform-events.js's 'submit form' event, it does get called, I see the log output, followed by:

```
Uncaught ReferenceError: Hooks is not defined
config.js?446624b38aaac055fc7ccdfd684b1d687cac71cc:12
AutoForm.addFormType.onSubmitautoform-events.js:311 autoFormSubmitHandlertemplate.js:463 (anonymous function)template.js:437 Template._withTemplateInstanceFunctemplate.js:462 (anonymous function)view.js:821 (anonymous function)view.js:475 Blaze._withCurrentViewview.js:820 (anonymous function)events.js:47 (anonymous function)jquery.js:4665 jQuery.event.dispatchjquery.js:4333 jQuery.event.add.elemData.handle
```

**Comment by abecks** [comment]
Ah yeah, my fix won't work out of scope of the package then. Can you guys try this:

* Remove my fix from your code
* Checkout the AutoForm repository to your project's ```packages``` folder as "aldeed_autoform"
* Restart your app

**Comment by abecks** [comment]
Actually, check out this one instead: https://github.com/abecks/meteor-autoform my fix is in place there.

**Comment by coniel** [comment]
That works.

**Comment by abecks** [comment]
Great, you can delete the folder once @aldeed merges in https://github.com/aldeed/meteor-autoform/pull/830

**Comment by coniel** [comment]
Ok, thanks a lot!

**Comment by garenor** [comment]
Thanks a lot!! It is working again!

**Comment by aldeed** [comment]
Released in 5.1.1, thanks

**Comment by fountainhead** [comment]
@garenor I am getting this same error in 5.1.3. The `onSubmit` hook does not get called. Is there something I am missing?


HTML Code is:

```html
<template name="subscriberNew">
  {{> quickForm collection="Subscribers" id="insertSubscriberForm" type="insert" fields="aNumber"}}
</template>
```


JS Code is:

```js
AutoForm.addHooks('insertSubscriberForm', {
    onSubmit: function (insertDoc, updateDoc, currentDoc) {
        this.event.preventDefault();
        console.log('before update call');
        return false;
    }
});
```

**Comment by coniel** [comment]
@fountainhead onSubmit is not called when using type="insert", only when no type is set or type is set to normal.

Check the docs:
Type "insert": https://github.com/aldeed/meteor-autoform#insert
Type "normal"/no type (which defaults to "normal"):  https://github.com/aldeed/meteor-autoform#normal
